e2e/test/scenarios/embedding/embedding-reproductions.cy.spec.js

```diff
@@ -1317,3 +1317,134 @@ describe("issue 50373", () => {
+      collectionIdAlias: "modelCollectionId",
+      cardIdAlias: "modelId",
+    });
+  });
+
+  it("should set the starting join step based on the query source", () => {
+    startNewEmbeddingQuestion();
+    const QA_DB_NAME = "QA Postgres12";
+    const DATA_SOURCE_NAME = "Orders";
+
+    cy.log("select a table as a data source");
+    H.popover().within(() => {
+      cy.findByText("Raw Data").click();
+      cy.findByRole("heading", { name: QA_DB_NAME }).click();
+      cy.findByRole("option", { name: DATA_SOURCE_NAME }).click();
+    });
+    H.getNotebookStep("data").button("Join data").click();
+
+    cy.log(
+      'select the "Join" step when the data source is a table will open a table in the same database',
+    );
+    H.popover().within(() => {
+      cy.findByText(QA_DB_NAME).should("be.visible");
+      cy.findByRole("option", { name: "Orders" }).should("be.visible");
+    });
+
+    cy.log(
+      "changing the data source while not selecting the join step should refresh the data picker on the join step",
+    );
+    H.getNotebookStep("data").findByText(DATA_SOURCE_NAME).click();
+
+    cy.log('go back to the "Bucket" step');
+    H.popover().within(() => {
+      cy.icon("chevronleft").click();
+      cy.icon("chevronleft").click();
+    });
+
+    cy.log(
+      "select a model as a data source should open the model step in the same collection as the data source",
+    );
+    H.popover().within(() => {
+      cy.findByText("Models").click();
+      cy.findByRole("menuitem", { name: COLLECTION_NAME }).click();
+      cy.findByRole("menuitem", { name: MODEL_IN_COLLECTION_NAME }).click();
+    });
+
+    H.popover().within(() => {
+      cy.log("the collection of the data source should be selected");
+      cy.findByRole("menuitem", { name: COLLECTION_NAME }).should(
+        "have.css",
+        "background-color",
+        // brand color
+        "rgb(80, 158, 227)",
+      );
+      cy.findByRole("menuitem", { name: MODEL_IN_COLLECTION_NAME })
+        .should("be.visible")
+        .click();
+    });
+
+    cy.log(
+      "select a data source after selecting a join step should refresh the data picker on the join step",
+    );
+    H.getNotebookStep("data").findByText(MODEL_IN_COLLECTION_NAME).click();
+    H.popover().within(() => {
+      cy.findByRole("menuitem", { name: "Our analytics" }).click();
+      cy.findByRole("menuitem", { name: MODEL_IN_ROOT_NAME }).click();
+    });
+
+    H.popover().within(() => {
+      cy.log("the collection of the new data source should be selected");
+      cy.findByRole("menuitem", { name: "Our analytics" }).should(
+        "have.css",
+        "background-color",
+        // brand color
+        "rgb(80, 158, 227)",
+      );
+      cy.findByRole("menuitem", { name: MODEL_IN_ROOT_NAME }).should(
+        "be.visible",
+      );
+    });
+  });
+
+  function startNewEmbeddingQuestion() {
+    cy.intercept("GET", "/api/search*", (req) => {
+      if (req.query.limit === "0") {
```

Comment:
I guess this is because the data picker search uses `/api/search` with limit="0"?

Reply:
yes, exactly. The embedding data picker doesn't care about the search items, it only needs to know how many tables and models there are.

Here's the call

https://github.com/metabase/metabase/blob/0c1a173665427d53152091c9213a493e50b559e8/frontend/src/metabase/querying/notebook/components/NotebookDataPicker/EmbeddingDataPicker/EmbeddingDataPicker.tsx#L28-L32

I couldn't find a way to just match the endpoint with the exact parameter, so I opted for global search match instead.